### PR TITLE
GHGA Validator

### PR DESCRIPTION
**Pull request by galinatremper**
GHGA Validator

### CODE REVIEW

File: .static_files_ignore [hunk 1]
```diff
@@ -1,2 +1,5 @@
 # Optional list of files which are actually static in the template
 # but are allowed to have different content in the current repository
+
+.github/workflows/check_config_docs.yaml
+scripts/license_checker.py
```
Comment (lkuchenb) [inline .static_files_ignore]:
```suggestion
```
You didn't (have to) change the license checker, right?

File: README.md [hunk 2]
```diff
@@ -1,106 +1,22 @@
...
 ```bash
-# Execute in the repo's root dir:
-pip install .
+# Install via pip
+pip install ghga-validator```
 
-# to run the service:
-my-microservice
+# Get help
+gdevutil --help
```
Comment (lkuchenb) [inline README.md]:
```suggestion
ghga-validator --help
```

File: .devcontainer/dev_install [hunk 3]
```diff
@@ -3,11 +3,13 @@
 
 cd /workspace
 
+git config --global --add safe.directory /workspace
+
```
Comment (lkuchenb) [inline .devcontainer/dev_install]:
```suggestion
```
Should not be part of dev_install

File: .devcontainer/dev_install [hunk 4]
```diff
@@ -3,11 +3,13 @@
...
 # upgrade pip
 python -m pip install --upgrade pip
 
 # install with all extras in editable mode
-pip install -e .[all]
+pip install -e .
```
Comment (lkuchenb) [inline .devcontainer/dev_install]:
```suggestion
pip install -e .
```
Why is this necessary?

File: ghga_validator/cli.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        report: The URL or path to store the validation results
+    """
+    with open(file, "r", encoding="utf8") as json_file:
+        submission_json = json.load(json_file)
+        if submission_json is None:
+            raise EOFError(f"<{file}> is empty! Nothing to validate!")
+        validation_report = validate(
+            str(Path(schema).resolve()),
+            submission_json,
+            "Submission",
```
Comment (lkuchenb) [inline ghga_validator/cli.py]: Please use the same pattern as used in the upstream validators, i.e. infer the root class from the model if a root is specified. We can offer an override CLI option just to be compatible with generic schemas, our immediate requirement is just a rooted schema.

File: ghga_validator/plugins/ref_validation.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+    ) -> ValidationMessage:
+        """
+        Get slot definition.
+
+        Args:
+            slot_name: slot name
+
+        Returns:
+            SlotDefinition: class definition
+
```
Comment (lkuchenb) [inline ghga_validator/plugins/ref_validation.py]: The docstring is unrelated

File: ghga_validator/plugins/ref_validation.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+            list_of_values = [value]
+        else:
+            list_of_values = value
+        if len(list_of_values) == 0:
+            return []
+        if not isinstance(list_of_values[0], dict):
+            return []
+        return list_of_values
+
+    def validate_json(
```
Comment (lkuchenb) [inline ghga_validator/plugins/ref_validation.py]: This function could use a more meaningful name. Also, it might be better to encapsulate the recursion into a class so that you don't have to carry over the constant arguments all the time and have a function with that many arguments

File: ghga_validator/plugins/ref_validation.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+        all_classes_name = self.schemaview.all_classes().keys()
+        if slot_def:
+            range_class = slot_def.range
+            if range_class in all_classes_name:
+                return range_class
+        return None
+
+    def to_list(self, value: Any) -> List[Dict]:
+        """
+        If an value is not an instance of List of objects, transform to it
```
Comment (lkuchenb) [inline ghga_validator/plugins/ref_validation.py]: This sounds generic, but there is additional logic in this function, i.e. it expects that the first list element is a dictionary?

File: ghga_validator/plugins/ref_validation.py [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+        Args:
+            obj: The object to validate
+            root: The root class name
+            class_name: The name of the class used in range
+            id_slot_name: The identifier slot in the range class
+
+        Returns:
+            List[str]: A list of all identifiers for the inlined objects
+
+        """
```
Comment (lkuchenb) [inline ghga_validator/plugins/ref_validation.py]: This function only works for top-level inlined objects. Also, it is called everytime you validate a reference and thus searches naively for every reference. Please refactor this so that the known references are iterated only once at the beginning of the validation and so that it works also for objects inlined anywhere in the data.

File: requirements-dev-common.txt [hunk 10]
```diff
@@ -28,3 +28,6 @@ mkdocs-material==9.0.3
 mkdocs-material-extensions==1.1.1
 mkdocstrings==0.19.1
 mkdocstrings-python-legacy==0.2.3
+
+linkml == 1.5.0
+linkml-validator==0.4.5
```
Comment (lkuchenb) [inline requirements-dev-common.txt]: These shouldn't be in the dev-common requirements but in the dev requirements.

File: ghga_validator/core/utils.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+def validate(
+    schema: str, obj: Dict, obj_type: str, plugins: List[Dict]
+) -> ValidationReport:
+    """
+    Validate an object of a particular type against a given schema.
+    Args:
+        schema: The URL or path to the schema YAML
+        obj: The object to validate
+        obj_type: The object type (schema type)
+    """
```
Comment (lkuchenb) [inline ghga_validator/core/utils.py]: It seems a bit odd that the top level function providing the core functionality of this package is in its `util` module

File: ghga_validator/plugins/ref_validation.py [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+        if "target_class" not in kwargs:
+            raise TypeError("'target_class' argument is required")
+        target_class = kwargs["target_class"]
+        messages = []
+        valid = True
+
+        messages = self.validate_json(obj, target_class, obj, target_class, "")
+        if len(messages) > 0:
+            valid = False
```
Comment (lkuchenb) [inline ghga_validator/plugins/ref_validation.py]:
```suggestion
        valid = len(messages) == 0:
```

File: ghga_validator/plugins/ref_validation.py [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+
+        Returns:
+            ValidationResult: A validation result that describes the outcome of validation
+
+        """
+        if "target_class" not in kwargs:
+            raise TypeError("'target_class' argument is required")
+        target_class = kwargs["target_class"]
+        messages = []
+        valid = True
```
Comment (lkuchenb) [inline ghga_validator/plugins/ref_validation.py]:
```suggestion
```

**Review by lkuchenb — CHANGES_REQUESTED** [review]
Thanks! Here are some general remarks in addition to the inline comments:

- [x] Please avoid static file changes unless absolutely necessary. I don't see why `dev_install` and `requirements-dev-common.txt` need changes for this tool
- [x] The devcontainer Docker Compose file can be reduced, we don't need the additional services for this CLI tool, also the environment variables are not used
- [ ] By default the CLI tool should output the errors in human readable form since we expect customers to use this tool, too. Let's brainstorm about that and make it a separate PR after this one, for now the JSON output file is fine.
- [x] Please add tests for the functionality and add a minimal schema / data with the test (not GHGA schema / data)
- [x] Align the report output format with the JSONSchemaValidator by Deepak (report the field with the erroneous value, not the alias, count the errors up, fix the list string in the output)

File: .devcontainer/dev_install (hunk 3 above)
Comment (galinatremper) [reply .devcontainer/dev_install]: Changed to template version of dev_install

File: .static_files_ignore (hunk 1 above)
Comment (galinatremper) [reply .static_files_ignore]: You're right!

File: README.md [hunk 14]
```diff
@@ -1,106 +1,22 @@
...
-# The entrypoint is pre-configured:
-docker run -p 8080:8080 ghga/my-microservice:<version>
-```
 
-If you prefer not to use containers, you may install the service from source:
 ```bash
-# Execute in the repo's root dir:
-pip install .
+# Install via pip
+pip install ghga_validator```
```
Comment (lkuchenb) [inline README.md]:
```suggestion
pip install ghga-validator```
```
PyPI normalizes the package name to the dash version. Both work, but the one with the dash is typically used when installing packages.

File: ghga_validator/cli.py (hunk 5 above)
Comment (galinatremper) [reply ghga_validator/cli.py]: I added target-class option to CLI (not required), if omitted, the root class is inferred.

File: ghga_validator/plugins/ref_validation.py (hunk 9 above)
Comment (galinatremper) [reply ghga_validator/plugins/ref_validation.py]: Refactored the implementation: now it is run only once and check all the levels of the JSON for the inlined objects.

File: ghga_validator/plugins/ref_validation.py (hunk 8 above)
Comment (galinatremper) [reply ghga_validator/plugins/ref_validation.py]: It can be list, but also a single value - I replaced it with the Union to make it more clear

File: ghga_validator/plugins/ref_validation.py (hunk 7 above)
Comment (galinatremper) [reply ghga_validator/plugins/ref_validation.py]: Renamed the function. Also the amount of arguments was automatically reduced after I moved the function for computing the list of ids outside the recursive function.

File: README.md [hunk 15]
```diff
@@ -1,106 +1,22 @@
...
-
----
-
-**\# please adapt the links of following badges:**
-![tests](https://github.com/ghga-de/my-microservice/actions/workflows/unit_and_int_tests.yaml/badge.svg)
-[![Coverage Status](https://coveralls.io/repos/github/ghga-de/my-microservice/badge.svg?branch=main)](https://coveralls.io/github/ghga-de/my-microservice?branch=main)
-# My-Microservice
-
-A description explaining the use case of this service.
+CLI Tool for validation of the (meta)data related to submissions.
```
Comment (lkuchenb) [inline README.md:3]:
```suggestion
CLI Tool for validation of GHGA metadata.
```

Comment (lkuchenb) [inline README.md]:
```suggestion
CLI Tool for validation of GHGA metadata.
```

**Review by sbilge — CHANGES_REQUESTED** [review]
I think it would be useful to provide one more example data (and a test) where it fails to pass the non-inlined reference validation as expected.

File: ghga_validator/plugins/ref_validation.py [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+
+        Args:
+            obj (Dict): The object to be parsed
+            target_class (str): Target class
+
+        Returns:
+            Dict[class_name, List[str]]: The dictionary containing the lists of
+            identifiers by the class name
+        """
+        all_ids = {}
```
Comment (lkuchenb) [inline ghga_validator/plugins/ref_validation.py]:
```suggestion
        all_ids = defaultdict(list)
```

File: ghga_validator/plugins/ref_validation.py [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+
+        for class_name, identifier, _, _ in ObjectIterator(
+            self.schemaview, obj, target_class
+        ):
+            if class_name not in all_ids:
+                all_ids[class_name] = [identifier]
+            else:
+                all_ids[class_name].append(identifier)
```
Comment (lkuchenb) [inline ghga_validator/plugins/ref_validation.py]:
```suggestion
            all_ids[class_name].append(identifier)
```

File: ghga_validator/plugins/ref_validation.py [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Plugin for LinkML JSON Validator used for validating the non inline references"""
+
+from typing import Dict, List, Optional, Union
```
Comment (lkuchenb) [inline ghga_validator/plugins/ref_validation.py:20]:
```suggestion
from collections import defaultdict
from numbers import Number
from typing import Dict, List, Optional, Union
```

File: ghga_validator/plugins/ref_validation.py [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+            kwargs: Additional arguments that are used for processing
+
+        Returns:
+            ValidationResult: A validation result that describes the outcome of validation
+
+        """
+        if "target_class" not in kwargs:
+            raise TypeError("'target_class' argument is required")
+        target_class = kwargs["target_class"]
+        messages = []
```
Comment (lkuchenb) [inline ghga_validator/plugins/ref_validation.py]:
```suggestion

```

File: ghga_validator/plugins/ref_validation.py [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+            List: List of missing references
+        """
+        if not isinstance(ref_value, list):
+            value_to_check = [ref_value]
+        else:
+            value_to_check = ref_value
+        non_match = []
+        if not all(x in id_list for x in value_to_check):
+            non_match = [x for x in value_to_check if x not in id_list]
+        return non_match
```
Comment (lkuchenb) [inline ghga_validator/plugins/ref_validation.py]:
```suggestion
        if not isinstance(ref_value, list):
            return [ref_value] if ref_value not in id_list else []
        return [x for x in ref_value if x not in id_list]
```

File: ghga_validator/plugins/ref_validation.py [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+                if range_class and not self.schemaview.is_inlined(slot_def):
+                    non_match = self.find_missing_refs(
+                        value, all_class_ids[range_class]
+                    )
+                    if len(non_match) == 0:
+                        continue
+                    message = ValidationMessage(
+                        severity=SeverityEnum.error,
+                        message="Unknown references "
+                        + f"({self.non_match_as_string(non_match)})",
```
Comment (lkuchenb) [inline ghga_validator/plugins/ref_validation.py]:
```suggestion
                        message="Unknown reference(s) " + str(non_match),
```
Let's keep it simple

File: ghga_validator/plugins/ref_validation.py [hunk 22]
```diff
@@ -0,0 +1,185 @@
...
+        Returns:
+            str: The validation error message
+        """
+        if len(non_match) > 1:
+            non_match_as_string = (
+                "'" + "', '".join(str(x) for x in non_match) + "'" + " were unexpected"
+            )
+        else:
+            non_match_as_string = "'" + non_match[0] + "'" + " was unexpected"
+        return non_match_as_string
```
Comment (lkuchenb) [inline ghga_validator/plugins/ref_validation.py]:
```suggestion
```

File: ghga_validator/plugins/ref_validation.py [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+
+        Returns:
+            Optional[str]: Range class for a slot
+        """
+        all_classes_name = self.schemaview.all_classes().keys()
+        if slot_def:
+            range_class = slot_def.range
+            if range_class in all_classes_name:
+                return range_class
+        return None
```
Comment (lkuchenb) [inline ghga_validator/plugins/ref_validation.py]:
```suggestion
        return (
            slot_def.range if slot_def.range in self.schemaview.all_classes() else None
        )
```
Why are you checking for the slot definition to cast to true? Any reason it may be None?

File: ghga_validator/plugins/ref_validation.py [hunk 24]
```diff
@@ -0,0 +1,185 @@
...
+                        message="Unknown references "
+                        + f"({self.non_match_as_string(non_match)})",
+                        field=f"{path_as_string(path)}.{field}",
+                        value=value,
+                    )
+                    messages.append(message)
+        return messages
+
+    def find_missing_refs(
+        self, ref_value: Union[List[str], str], id_list: List
```
Comment (lkuchenb) [inline ghga_validator/plugins/ref_validation.py]:
```suggestion
        self, ref_value: Union[List[Union[Number, str]], Union[Number,str]], id_list: List
```
All types in LinkML can be identifiers

**Review by lkuchenb — CHANGES_REQUESTED** [review]

File: ghga_validator/plugins/ref_validation.py [hunk 25]
```diff
@@ -0,0 +1,185 @@
...
+        Args:
+            obj: The object to validate
+            kwargs: Additional arguments that are used for processing
+
+        Returns:
+            ValidationResult: A validation result that describes the outcome of validation
+
+        """
+        if "target_class" not in kwargs:
+            raise TypeError("'target_class' argument is required")
```
Comment (lkuchenb) [inline ghga_validator/plugins/ref_validation.py]: Since you're writing the surrounding validator tool you are in control whether this argument is passed or not. What is the advantage of catching the `KeyError` and raising a `TypeError` instead?

**Review by lkuchenb — CHANGES_REQUESTED** [review]

File: ghga_validator/plugins/ref_validation.py (hunk 23 above)
Comment (galinatremper) [reply ghga_validator/plugins/ref_validation.py]: You're right - when we using get_induced_slot, it cannot be None, because in this case we would receive the ValueError.

File: ghga_validator/plugins/ref_validation.py (hunk 24 above)
Comment (galinatremper) [reply ghga_validator/plugins/ref_validation.py]: Agree, string is used as identifier type for the submission, but it should of course have different type in the generic implementation.

File: ghga_validator/cli.py [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+import json
+from pathlib import Path
+from typing import Optional
+
+import typer
+from linkml_validator.plugins.jsonschema_validation import JsonSchemaValidationPlugin
+
+from ghga_validator.core.validation import get_target_class, validate
+
+from .plugins.ref_validation import RefValidationPlugin
```
Comment (lkuchenb) [inline ghga_validator/cli.py]:
```suggestion
from ghga_validator.plugins import RefValidationPlugin
```
One more thing: Please create an `__init__.py` in the plugins module and import all plugins there.

**Review by lkuchenb — CHANGES_REQUESTED** [review]

File: ghga_validator/cli.py [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+        if not validation_report.valid:
+            with open(report, "w", encoding="utf-8") as sub:
+                json.dump(
+                    validation_report.dict(exclude_unset=True, exclude_none=True),
+                    sub,
+                    ensure_ascii=False,
+                    indent=4,
+                )
+            return False
+    return True
```
Comment (lkuchenb) [inline ghga_validator/cli.py]:
```suggestion
    if submission_json is None:
        raise EOFError(f"<{file}> is empty! Nothing to validate!")
    validation_report = validate(
        str(Path(schema).resolve()),
        target_class=target_class,
        obj=submission_json,
        plugins=VALIDATION_PLUGINS,
    )
    with open(report, "w", encoding="utf-8") as sub:
        json.dump(
            validation_report.dict(exclude_unset=True, exclude_none=True),
            sub,
            ensure_ascii=False,
            indent=4,
        )
    return validation_report.valid
```

And another comment - the validation report should always be written, also if it's positive. Also avoid open file descriptors.

**Review by lkuchenb — CHANGES_REQUESTED** [review]

**Comment by galinatremper** [comment]
> I think it would be useful to provide one more example data (and a test) where it fails to pass the non-inlined reference validation as expected.

More test cases will be added in a separate PR.

**Review by lkuchenb — APPROVED** [review]

**Review by sbilge — APPROVED** [review]
